connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala

```diff
@@ -241,6 +241,91 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
     assert(rows10(0).getString(0) === "amy")
     assert(rows10(1).getString(0) === "alex")
   }
+
+  test("SPARK-50793: MySQL JDBC Connector failed to cast some types") {
+    val tableName = catalogName + ".test_cast_function"
+    withTable(tableName) {
+      val stringValue = "0"
+      val stringLiteral = "'0'"
+      val longValue = 0L
+      val binaryValue = Array[Byte](0x30)
+      val binaryLiteral = "x'30'"
+      val doubleValue = 0.0
+      val doubleLiteral = "0.0"
+      // CREATE table to use types defined in Spark SQL
+      sql(
+        s"CREATE TABLE $tableName (string_col STRING, long_col LONG, " +
+          "binary_col BINARY, double_col DOUBLE)")
+      sql(
+        s"INSERT INTO $tableName VALUES($stringLiteral, $longValue, $binaryLiteral, $doubleValue)")
+
+      def testCast(
+          castType: String,
+          sourceCol: String,
+          targetCol: String,
+          sourceDataType: DataType,
+          sourceValue: Any,
+          targetDataType: DataType,
+          targetValue: Any): Unit = {
+        val sql = s"SELECT $sourceCol AS source, CAST($sourceCol AS $castType) AS target " +
+          s"FROM $tableName WHERE CAST($sourceCol AS $castType) = $targetCol"
```

Comment:
Shall we only test the `CAST($sourceCol AS $castType)`? It seems no need to test `$sourceCol AS source`.

Reply:
Hello @beliefer could you please help me confirm if we should remove `$sourceCol as source`, thanks.

Reply:
```
val sql = s"SELECT CAST($sourceCol AS $castType) AS target " +
          s"FROM $tableName WHERE CAST($sourceCol AS $castType) = $targetCol"
```
You can simplify the test SQL. No need to check

Reply:
Sure, they are removed. PTAL, thanks.

Reply:
I have an idea that no matter the collation is, we convert them to the default collation.
`For example, SELECT CAST($sourceCol AS $castType) collate 'ucs_basic'.`